src/shims/posix/linux/sync.rs

```diff
@@ -0,0 +1,88 @@
+use crate::*;
+use rustc_target::abi::{Align, Size};
+
+/// Implementation of the SYS_futex syscall.
+pub fn futex<'tcx>(
+    this: &mut MiriEvalContext<'_, 'tcx>,
+    args: &[OpTy<'tcx, Tag>],
+    dest: PlaceTy<'tcx, Tag>,
+) -> InterpResult<'tcx> {
+    // The amount of arguments used depends on the type of futex operation.
+    // Some users always pass all arguments, even the unused ones, due to how they wrap this syscall in their code base.
+    // Some other users pass only the arguments the operation actually needs. So we don't use `check_arg_count` here.
+    if !(4..=7).contains(&args.len()) {
```

Comment:
The argument should not be about what users do, the argument should be about what is allowed by the API and calling convention. If users cause UB by what they do, we should still reject such code. ;)

Reply:
That makes sense. Though, what users *do* is an important part in the decision to accept different number of arguments here. For example, Miri's implementation of `SYS_getrandom` only accepts exactly three arguments, even though it'd be perfectly safe to pass 6. But that's not a problem, because nobody ever calls getrandom with six arguments. The only thing different for SYS_futex is that people actually do call it with extra unused arguments. (The example in the man page even does so.)

Anyway, updated the comment to focus more on the perspective of the api instead of the user.

Reply:
> For example, Miri's implementation of SYS_getrandom only accepts exactly three arguments, even though it'd be perfectly safe to pass 6.

That is just based on me not knowing that it is okay to pass more arguments. Assuming that this really is true for all calling conventions, we should relax the check for getrandom as well.

Does it matter which types those extra arguments have? Like, would it also be okay to pass some huge struct by-value, or does it have to be something that fits into a register?

Reply:
Uhh, yeah those are good questions. I think that'd be okay, but I'd have to go digging in all the calling conventions to verify. The libc `syscall()` function is declared in C (and Rust) as a variadic function, but is implemented for all platforms in assembly assuming always exactly 7 or 8 arguments (depending on the platform). The implementation always copies all these arguments over to the right registers and then triggers the `syscall` instruction. So any arguments rhat did not end up in the places where a set of regular `(int, int, int, int, int, int, int)` arguments to a variadic function would go, are simply lost but don't affect the syscall.

Reply:
Okay, so maybe leave a FIXME?
I guess we'll need a special thing for syscalls that replaces the usual argument check with "make sure all arguments have scalar layout and that there are not too many of them".